#!/bin/bash

###########################################
## Adapter trimming for raw data genomes ##
###########################################

# This script will remove adapters and low quality sequences from our raw data genomes, sequenced in Illumina 1.3+ platforms. 
# The script must take 2.5-4 hours to run depending on library size (5X-24X).

###################################
## Variable and PATHs definition ##
###################################
# We will have five PATHs:

# PATH to all intermediate (.bam) files of this pipeline
BAMFOLD='/path/to/bamfolder'
# PATH to all sanity checks (flagstat files)
FLAGFOLD='/path/to/flagstat/output'
# PATH for temporary files generated by bwa mem
TEMP='/path/to/store/temporary/files'
# PATH where our program executables are stored (can also be defined as variables)
PROG='/path/to/programs/folder'
# PATH with the folder where we downloaded our raw genome data.
GENFOLD='/path/to/genomes'

#################################
## Sequence quality assessment ##
#################################

# First we want to estimate and visualize per base sequence quality, k-mer and GC content, sequence length distribution...
# We will do this with FastQC: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/
# This example was done with a library of the Algerian African golden wolf - Liu et al., (2018)

$PROG/FastQC/fastqc $GENFOLD/C4PCGANXX_6.R1.fastq.gz -o $BAMFOLD
$PROG/FastQC/fastqc $GENFOLD/C4PCGANXX_6.R2.fastq.gz -o $BAMFOLD

######################
## Adapter trimming ##
######################

# We will remove adapters commonly used in Illumina 1.3+ platforms. 

python $PROG/cutadapt/bin/cutadapt -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT --minimum-length 20 -q 20 -o $BAMFOLD/C4PCGANXX_6.trimmed.R1.fastq.gz -p $BAMFOLD/C4PCGANXX_6.trimmed.R2.fastq.gz $GENFOLD/C4PCGANXX_6.R1.fastq.gz $GENFOLD/C4PCGANXX_6.R2.fastq.gz 2> $FLAGFOLD/reportcut.afr_wolf.algeria.txt

#################################
## Sequence quality assessment ##
#################################

# We want to check how if our adapters have been correctly trimmed. We run FastQC again. 
$PROG/FastQC/fastqc $BAMFOLD/C4PCGANXX_6.trimmed.R1.fastq.gz -o $BAMFOLD
$PROG/FastQC/fastqc $BAMFOLD/C4PCGANXX_6.trimmed.R2.fastq.gz -o $BAMFOLD

################
## REFERENCES ##
################

# Liu, Y. H., et al. (2018). Whole-Genome sequencing of African dogs provides insights into adaptations against tropical parasites. Molecular Biology and Evolution, 35(2), 287â€“298. https://doi.org/10.1093/molbev/msx258
